src/omp/backend.f90

```diff
@@ -283,24 +292,29 @@ subroutine tds_solve_dist(self, du, u, dirps, tdsops)
     class(field_t), intent(in) :: u
     type(dirps_t), intent(in) :: dirps
     class(tdsops_t), intent(in) :: tdsops
-    integer :: n_halo
+    integer :: n_halo, n, n_groups, dir
 
     ! TODO: don't hardcode n_halo
     n_halo = 4
+    dir = u%dir
+    n = self%mesh%get_n(u)
```

Comment:
yeah, that's my issue with `NONE`, by allowing it it means that we can never use `u%data_loc` due to the off chance it can be `NONE`. I feel like we should keep `NONE` only for edge cases or test etc. but not allow it in these kind of portion of the code.

I haven't added it yet, but I am planning on having the code through an error and crash when using `NONE` in not allowed places.

Reply:
This links back to `field%n` and why I was really concerned about it. I have the exact same concerns using `get_n` here. Having the `data_loc` in `field` optional makes it impossible to use `get_n` in some places, and I think this is a very good feature, and that's why I'm completely happy if `data_loc` is optional, but really concerned otherwise. And requiring a `data_loc` in most cases is effectively equivalent to making it a normal argument, not an optional one. Regarding having a `data_loc` in a field, I think just the opposite, and we should use `NONE` in most cases and a `VERT` or similar in rare cases. Because otherwise we're making the solver unnecessarily complicated. There are also cases where we can't specify a `data_loc`, but still need to use `tds_solve`.

Can you share your opinion on using `tdsops_t%n` here instead? A `tdsops` object is the building block of a `tds_solve` call and we rely on a lot of data it provides to us, including the `n` of the resulting field. In my opinion `get_n` is a good replacement here over `tdsops%n`.

Reply:
do you have examples where specifying the `data_loc` in the field won't work?

I am having issues understanding where/why it might be an issue. At the end of the day, a field has some data in it located on either vertices, cell centers, face etc. and that is 'constant' throughout its life. If the tds solve needs to loop or do something that is not of the size of the data it is working with, maybe that just needs to be handled by tds solve itself but won't affect `n`. 
What I mean is that `get_n` will always give you the `n` of the data it stores, but if for some reason tds_solve needs `n+1` or something else, then this needs to be handled differently. And it shouldn't be called `n`.

My issue with using `tdsops_t%n` is that it makes everything less clear, we will have two different `n` and depending on what we need to do we use one or the other. That's confusing for anyone (like myself) not very familiar with the tdsops module.

Reply:
First a very quick summary of `tds_solve` design logic: tridiagonal solver algorithm implementation (in distributed.f90) uses preprocessed set of arrays all sized equivalent to the size of tridiagonal system they solve, and lets call this size `n_tds` for the moment for clarification (this is stored in `tdsops%n`). Then there are loops in the implementation where we traverse the domain as well as these preprocessed arrays. These loops ranges are equivalent to the size of the tridiagonal system we're solving and I think its sensible to define the loop ranges based on `n_tds`. The meaningful number of entries in the output field is practically determined by the `n_tds`, so making sure that we have the right `n` in the output field beforehand and using this `n` in the loop ranges isn't a straightforward process. While the `tdsops` operator knows the size of the tridiagonal system it solves, determining the loop range in a different way from a different object is not a good idea in my opinion, and I think its error prone.

Also, replacing `tdsops%n` here with `get_n` necessitates us to specify the `data_loc` of every single field we request, even if they're short lived intermediate storages. Its true that a field stores data in one of these predetermined locations, but ambiguity around this is helping us with abstraction. For example we can pass two fields with `data_loc` `NONE`, and still get a `tds_solve` work correctly, and would be nice to keep this. 

One example where we can use this ambiguity and thus having a `data_loc` won't work:
Lets say we have a scalar at `CELL` and want to obtain say the x-derivative at `VERT`. We need 3 `tds_solve`s to get to the final result, but there are multiple paths we can choose. For example we can do `CELL` -> `Z_FACE` -> `X_EDGE` -> `VERT` or `CELL` -> `Z_FACE` -> `Y_EDGE` -> `VERT`. Its possible for us to carry out both these operations with a single subroutine such as `call custompath_p2v(dpdx, p, tdsopsA, tdsopsB, tdsopsC)`. Whatever path we decide to carry out is basically specified when we call this subroutine, and because they pass through different `data_loc`s, we can't specify the `data_loc` of intermediate fields inside the subroutine by hand. However, as long as we pass the correct set of `tdsopsA/B/C` operators, we're certain that the path we specify is respected and we obtain the correct result at the end.

I see that the naming in `tdsops%n` can be confusing, so like I used above maybe renaming it to `tdsops%n_tds` or anything else will distinguish it from `n` and make things more clear. Also, its use is strictly in the context of `tds_solve` functions I think, and really shouldn't be used anywhere else. I'll be happy to add documentation regarding `tds_solve` stuff if you think it will be helpful. What do you think?

Reply:
Thanks for the detailed explanation, it is a lot more clear now. So let's do that, rename `tdsops%n` to `tdsops%n_tds` and use it when we have to, but keep the possibility to have `data_loc` specified when we want `n` outside the solver.

As a general rule, I think it will still be good to define `data_loc` whenever possible, even if not used directly it will help debug things and check for consistency of sizes at times.

And indeed, having these kind of high level design choices written somewhere could be useful, specially for anyone getting started with contributing to x3d2.